instrumentation/opentelemetry-instrumentation-urllib/src/opentelemetry/instrumentation/urllib/__init__.py

```diff
@@ -0,0 +1,233 @@
+    MetricMixin,
+)
+from opentelemetry.instrumentation.urllib.version import __version__
+from opentelemetry.instrumentation.utils import http_status_to_status_code
+from opentelemetry.trace import SpanKind, get_tracer
+from opentelemetry.trace.status import Status, StatusCode
+
+# A key to a context variable to avoid creating duplicate spans when instrumenting
+_SUPPRESS_URLLIB_INSTRUMENTATION_KEY = "suppress_urllib_instrumentation"
+
+
+# pylint: disable=unused-argument
+# pylint: disable=R0915
+def get_default_span_name(method):
+    """Default implementation for name_callback, returns HTTP {method_name}."""
+    return "HTTP {}".format(method).strip()
+
+
+def _instrument(tracer_provider=None, span_callback=None, name_callback=None):
+    """Enables tracing of all requests calls that go through
+    :code:`urllib.Client._make_request`"""
+
+    opener_open = OpenerDirector.open
+
+    @functools.wraps(opener_open)
+    def instrumented_request(opener, fullurl, data=None, timeout=None):
+
+        if isinstance(fullurl, str):
+            request_ = Request(fullurl, data)
+        else:
+            request_ = fullurl
+
+        def get_or_create_headers():
+            return getattr(request_, "headers", {})
+
+        def call_wrapped():
+            return opener_open(opener, request_, data=data, timeout=timeout)
+
+        return _instrumented_requests_call(
+            opener, request_, call_wrapped, get_or_create_headers
+        )
+
+    def _instrumented_requests_call(
+        opener, request, call_wrapped, get_or_create_headers
+    ):  # pylint: disable=too-many-locals
+        if context.get_value("suppress_instrumentation") or context.get_value(
+            _SUPPRESS_URLLIB_INSTRUMENTATION_KEY
+        ):
+            return call_wrapped()
+
+        method = request.get_method().upper()
+        url = request.full_url
+
+        span_name = ""
+        if name_callback is not None:
+            span_name = name_callback()
+        if not span_name or not isinstance(span_name, str):
+            span_name = get_default_span_name(method)
+
+        recorder = URLLibInstrumentor().metric_recorder
+
+        labels = {
+            "http.method": method,
+            "http.url": url,
+        }
+
+        with get_tracer(
+            __name__, __version__, tracer_provider
+        ).start_as_current_span(span_name, kind=SpanKind.CLIENT) as span:
+            exception = None
+            with recorder.record_client_duration(labels):
+                if span.is_recording():
+                    span.set_attribute("component", "http")
+                    span.set_attribute("http.method", method)
+                    span.set_attribute("http.url", url)
+
+                headers = get_or_create_headers()
+                propagators.inject(type(headers).__setitem__, headers)
+
+                token = context.attach(
+                    context.set_value(
+                        _SUPPRESS_URLLIB_INSTRUMENTATION_KEY, True
```

Comment:
can you clarify why this is needed? It seems to me that, as this is the request being sent, it would be impossible really to have subsequent logic also fire another span, or do something that would create such spans.

Reply:
That's right, but here we only wrap this request or I've missing something ?

Reply:
I believe we do this for [requests](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/master/instrumentation/opentelemetry-instrumentation-requests/src/opentelemetry/instrumentation/requests/__init__.py#L58) because some of the request calls have the same code path (but we instrument all of them) so we add this context variable to avoid duplicates. I'm not sure if urllib has the same issue though.

Reply:
We had a situation where this was needed in the `botocore` instrumentation to avoid infinite recursion. You can use `botocore` to send traces to our `X-Ray` backend using `botocore.client('x-ray').put_trace_segments()`. We had created our own _exporter_ for the traces to not use the OTel Collector. So we would have:

1. Send trace using our exporter and `botocore.put_trace_segments()`
2. `BotocoreInstrumentor` intercepts call to `botocore` and generates a new trace A.
3. Send trace A using our exporter and `botocore.put_trace_segments()` to X-Ray.
4. `BotocoreInstrumentor` intercepts call to `botocore` and generates a new trace B.
5. Send trace B using our exporter and `botocore.put_trace_segments()` to X-Ray.

etc.

I imagine if someone exports using `urllib` they would encounter a similar problem.

Reply:
That makes sense. But isn't the solution in that scenario to have the exporter and instrumentation honor the "suppress_instrumentation" value? 

@NathanielRN in the scenario you stated, the `put_trace_segements` in 1 should not have emitted a trace, avoiding this whole loop. If the `BotocoreInstrumentor` checks for the "suppress_instrumentation" context value, it should have emitted nothing.